File: Heat/Authorize/ResourceClaimsAuthorizeManager.vb
Imports System.Security.Claims
Imports Heat.Repositories

Public Class ResourceClaimsAuthorizeManager
    Inherits ClaimsAuthorizationManager

    Private _db As HeatDBContext

    Sub New()
    End Sub

    Sub New(dbContext As HeatDBContext)
        _db = dbContext
    End Sub

    Public Overrides Function CheckAccess(context As AuthorizationContext) As Boolean
        Dim resource = context.Resource.First
        Dim claims = context.Principal.Claims.ToList
        Dim action = context.Action.First

        'http://stackoverflow.com/questions/26464848/custom-authorization-in-asp-net-webapi-what-a-mess
        Dim authRequired As ResourceOperations

        authRequired = [Enum].Parse(GetType(ResourceOperations), action.Value)

        For Each Claim In claims
            If Claim.Type = "http://scheme.diegobrocchi.it/claims/2015/operation" Then
                Dim authOwned As ResourceOperations = [Enum].Parse(GetType(ResourceOperations), Claim.Value)
                If authOwned And authRequired = authRequired Then
                    Return True
                End If
            End If
        Next

        Return False
    End Function



End Class

File: Heat.ConvertedToC#/Authorize/ResourceClaimsAuthorizeManager.cs
using System;
using System.Linq;
using System.Security.Claims;
using Heat.Repositories;
namespace Heat
{

    public class ResourceClaimsAuthorizeManager : ClaimsAuthorizationManager
	{


		private HeatDBContext _db;
		public ResourceClaimsAuthorizeManager()
		{
		}

		public ResourceClaimsAuthorizeManager(HeatDBContext dbContext)
		{
			_db = dbContext;
		}

		public override bool CheckAccess(System.Security.Claims.AuthorizationContext context)
		{
			var resource = context.Resource.First();
			var claims = context.Principal.Claims.ToList();
			var action = context.Action.First();

			//http://stackoverflow.com/questions/26464848/custom-authorization-in-asp-net-webapi-what-a-mess
			ResourceOperations authRequired = default(ResourceOperations);

			authRequired =(ResourceOperations) Enum.Parse(typeof(ResourceOperations), action.Value);

			foreach (Claim singleClaim in claims) {
				 
				if (singleClaim.Type == "http://scheme.diegobrocchi.it/claims/2015/operation")
                {
					ResourceOperations authOwned = (ResourceOperations) Enum.Parse(typeof(ResourceOperations), singleClaim.Value);
					if ((authOwned & authRequired) == authRequired) {
						return true;
					}
				}
			}

			return false;
		}



	}
}
